[APF-14][APPS-1875] Add Private Action Runner helm chart (#1465)
DataDog/helm-charts: charts/private-action-runner/README.md

This Helm Chart will deploy the Datadog Private Action runner inside a Kubernetes cluster.
You will be able to use private actions from the Datadog Workflow and Datadog App Builder products.
When deploying this chart, you will be able to give permissions to the runner in order to be able to run Kubernetes actions.

## Requirements
* A Datadog account with private actions enabled
* The kubectl cli
* Helm
* Sufficient permissions to the kubernetes cluster

## In order to use this chart

* Go to the private action runner tab https://app.datadoghq.com/workflow/private-action-runners
* Create a new private action runner
* Follow the instructions and you should have a running docker container and `config/config.yaml` file
* Stop the docker container (`docker stop <name-of-the-container>` or `docker compose stop`)
* Replace the `URN_FROM_CONFIG` and the `PRIVATE_KEY_FROM_CONFIG` from the chart's `values.yaml` with the `urn` and the `privateKey` from the `config/config.yaml` of the docker container.
* Create the kubernetes namespace for your runner `kubectl create namespace private-action-runner`
* Replace the `namespace` in the `values.yaml` with the namespace you just created

> * Install the Helm chart `helm install <name> .`

Reviewer: I think we can reference the chart public URL ? But customers should have other instructions on how to pass the values

Author: I saw synthetics has https://github.com/DataDog/helm-charts/blob/bc1b5bb922e68106f1b6e222bd2b7e90a90d351b/charts/synthetics-private-location/README.md.gotmpl#L11C1-L23C1
so I reworked our instructions to be more similar to theirs wdyt?

Author: agreed, I renamed it to `config.yaml`

# Private action runner Helm chart

> This Helm Chart will deploy the Datadog Private Action runner inside a Kubernetes cluster.
> You will be able to use private actions from the Datadog Workflow and Datadog App Builder products.
> When deploying this chart, you will be able to give permissions to the runner in order to be able to run Kubernetes actions.

Reviewer 2: ```suggestion
This Helm Chart deploys the Datadog Private Action runner inside a Kubernetes cluster. It allows you to use private actions from the Datadog Workflow and Datadog App Builder products. When deploying this chart, you can give permissions to the runner in order to be able to run Kubernetes actions.
```

## Requirements
* A Datadog account with private actions enabled

> * The kubectl cli
> * Helm
> * Sufficient permissions to the kubernetes cluster

Reviewer 2: ```suggestion
* The `kubectl` cli
* Helm
* Sufficient permissions to the Kubernetes cluster
```

> ## In order to use this chart

Reviewer 2: ```suggestion
## Use this chart
```

> * The kubectl cli

Reviewer 2: ```suggestion
* The `kubectl` cli
```

> * Go to the private action runner tab https://app.datadoghq.com/workflow/private-action-runners
> * Create a new private action runner
> * Follow the instructions and you should have a running docker container and `config/config.yaml` file
> * Stop the docker container (`docker stop <name-of-the-container>` or `docker compose stop`)
> * You must create a `config.yaml` file with the appropriate values. An example `config.yaml` file is provided in the `examples` directory for you to copy.
>     * You must replace the `URN_FROM_CONFIG` and the `PRIVATE_KEY_FROM_CONFIG` in the example file with with the `urn` and the `privateKey` from the `config/config.yaml` of the docker container
>     * Other values can be re-configured as well, but you can also choose to stick with the defaults in the example
> * You need to add this repository to your Helm repositories:
>     ```
>     helm repo add datadog https://helm.datadoghq.com
>     helm repo update
>     ```
> * Install the Helm chart:
>     ```bash
>         helm install <RELEASE_NAME> datadog/private-action-runner -f ./config.yaml
>     ```
> * Go to the workflow connections https://app.datadoghq.com/workflow/connections
> * Create a new connection, select your private action runner and use `Service account authentication`
> * Create a new workflow and use a kubernetes action like `List pod` / `List deployment`
> 
> ## Going further
> * Adjust the service account permissions according to your needs, you can find more informations about kubernetes RBAC here https://kubernetes.io/docs/reference/access-authn-authz/rbac
> * Deploy several runners with different permissions / create different connections according to your needs
> * Private actions documentation https://docs.datadoghq.com/service_management/app_builder/private_actions

Reviewer 2: ```suggestion
1. Go to the [private action runner tab](https://app.datadoghq.com/workflow/private-action-runners).
2. Create a new private action runner.
3. Follow the instructions. You now have a running docker container and `config/config.yaml` file.
4. Stop the docker container (`docker stop <name-of-the-container>` or `docker compose stop`).
5. Create a `config.yaml` file with the appropriate values. An example `config.yaml` file is provided in the `examples` directory for you to copy.
    * Replace the `URN_FROM_CONFIG` and the `PRIVATE_KEY_FROM_CONFIG` in the example file with with the `urn` and the `privateKey` from the `config/config.yaml` of the docker container.
    * You can reconfigure other values or use the defaults in the example.
6. Add this repository to your Helm repositories:
    ```
    helm repo add datadog https://helm.datadoghq.com
    helm repo update
    ```
7. Install the Helm chart:
    ```bash
        helm install <RELEASE_NAME> datadog/private-action-runner -f ./config.yaml
    ```
8. Go to the [Workflow connections page](https://app.datadoghq.com/workflow/connections).
9. Create a new connection, select your private action runner, and use **Service account authentication**.
10. Create a new workflow and use a Kubernetes action like **List pod** or **List deployment**.

## Going further
* Adjust the service account permissions according to your needs. Learn more about [Kubernetes RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac).
* Deploy several runners with different permissions or create different connections according to your needs.
* Learn more about [Private actions](https://docs.datadoghq.com/service_management/app_builder/private_actions).
```

## Values

> | Key | Type | Default | Description |
> |-----|------|---------|-------------|
> | common.image | string | `"us-east4-docker.pkg.dev/datadog-sandbox/apps-on-prem/onprem-runner:v0.0.1-alpha23"` | Current Datadog Private Action Runner image |
> | runners[0].config | object | `{"actionsAllowlist":["com.datadoghq.kubernetes.core.listPod"],"appBuilder":{"port":9016},"ddBaseURL":"https://app.datadoghq.com","modes":["workflowAutomation","appBuilder"],"privateKey":"PRIVATE_KEY_FROM_CONFIG","urn":"URN_FROM_CONFIG"}` | Configuration for the Datadog Private Action Runner |
> | runners[0].config.actionsAllowlist | list | `["com.datadoghq.kubernetes.core.listPod"]` | List of actions that the Datadog Private Action Runner is allowed to execute |
> | runners[0].config.appBuilder.port | int | `9016` | Required port for App Builder Mode |
> | runners[0].config.ddBaseURL | string | `"https://app.datadoghq.com"` | Base URL of the Datadog app |
> | runners[0].config.modes | list | `["workflowAutomation","appBuilder"]` | Modes that the runner can run in |
> | runners[0].config.privateKey | string | `"PRIVATE_KEY_FROM_CONFIG"` | The runner's privateKey from the enrollment page |
> | runners[0].config.urn | string | `"URN_FROM_CONFIG"` | The runner's URN from the enrollment page |
> | runners[0].kubernetesPermissions | list | `[{"apiGroups":[""],"resources":["pods"],"verbs":["list","get"]},{"apiGroups":["apps"],"resources":["deployments"],"verbs":["list","get"]}]` | List of Kubernetes permissions that the Datadog Private Action Runner has |
> | runners[0].name | string | `"default"` | Name of the Datadog Private Action Runner |
> | runners[0].replicas | int | `1` | Number of pod instances for the Datadog Private Action Runner |

Reviewer 2: ```suggestion
| Key                                | Description                                                                  | Type   | Default                                                                                                                                                                                                                                        |
|------------------------------------|------------------------------------------------------------------------------| -------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| common.image                       | Current Datadog Private Action Runner image                                  | string | `"us-east4-docker.pkg.dev/datadog-sandbox/apps-on-prem/onprem-runner:v0.0.1-alpha23"`                                                                                                                                                          |
| runners[0].config                  | Configuration for the Datadog Private Action Runner                          | object | `{"actionsAllowlist":["com.datadoghq.kubernetes.core.listPod"],"appBuilder":{"port":9016},"ddBaseURL":"https://app.datadoghq.com","modes":["workflowAutomation","appBuilder"],"privateKey":"PRIVATE_KEY_FROM_CONFIG","urn":"URN_FROM_CONFIG"}` |
| runners[0].config.actionsAllowlist | List of actions that the Datadog Private Action Runner is allowed to execute | list   | `["com.datadoghq.kubernetes.core.listPod"]`                                                                                                                                                                                                    |
| runners[0].config.appBuilder.port  | Required port for App Builder Mode                                           | int    | `9016`                                                                                                                                                                                                                                         |
| runners[0].config.ddBaseURL        | Base URL of the Datadog app                                                  | string | `"https://app.datadoghq.com"`                                                                                                                                                                                                                  |
| runners[0].config.modes            | Modes that the runner can run in                                             | list   | `["workflowAutomation","appBuilder"]`                                                                                                                                                                                                          |
| runners[0].config.privateKey       | The runner's privateKey from the enrollment page                             | string | `"PRIVATE_KEY_FROM_CONFIG"`                                                                                                                                                                                                                    |
| runners[0].config.urn              | The runner's URN from the enrollment page                                    | string | `"URN_FROM_CONFIG"`                                                                                                                                                                                                                            |
| runners[0].kubernetesPermissions   | List of Kubernetes permissions that the Datadog Private Action Runner has    | list   | `[{"apiGroups":[""],"resources":["pods"],"verbs":["list","get"]},{"apiGroups":["apps"],"resources":["deployments"],"verbs":["list","get"]}]`                                                                                                   |
| runners[0].name                    | Name of the Datadog Private Action Runner                                    | string | `"default"`                                                                                                                                                                                                                                    |
| runners[0].replicas                | Number of pod instances for the Datadog Private Action Runner                | int    | `1`                                                                                                                                                                                                                                            |
```

Author: @neko-dd this section of the README is auto-generated from the comments in values.yaml. If I change it directly then it will cause tests to fail. It looks like these are formatting changes? so I don't think I am able to change this

Reviewer 2: I see. There's no way to reorder the columns, I take it? That's mostly what I was trying to do.

Author: yeah I don't think so unfortunately

Reviewer 2: Okay. Alas!

> | Key | Type | Default | Description |
> |-----|------|---------|-------------|
> | common.image | string | `"us-east4-docker.pkg.dev/datadog-sandbox/apps-on-prem/onprem-runner:v0.0.1-alpha22"` | Current Datadog Private Action Runner image |
> | runners[0] | object | `{"config":{"actionsAllowlist":["com.datadoghq.kubernetes.core.listPod"],"appBuilder":{"port":9016},"ddBaseURL":"https://app.datadoghq.com","modes":["workflowAutomation","appBuilder"],"privateKey":"PRIVATE_KEY_FROM_CONFIG","urn":"URN_FROM_CONFIG"},"kubernetesPermissions":[{"apiGroups":[""],"resources":["pods"],"verbs":["list","get"]},{"apiGroups":["apps"],"resources":["deployments"],"verbs":["list","get"]}],"name":"default","replicas":1}` | Name of the Datadog Private Action Runner |
> | runners[0].config | object | `{"actionsAllowlist":["com.datadoghq.kubernetes.core.listPod"],"appBuilder":{"port":9016},"ddBaseURL":"https://app.datadoghq.com","modes":["workflowAutomation","appBuilder"],"privateKey":"PRIVATE_KEY_FROM_CONFIG","urn":"URN_FROM_CONFIG"}` | This is the configuration for the Datadog Private Action Runner |
> | runners[0].config.actionsAllowlist | list | `["com.datadoghq.kubernetes.core.listPod"]` | List of actions that the Datadog Private Action Runner is allowed to execute |
> | runners[0].config.appBuilder.port | int | `9016` | Required port for App Builder Mode |
> | runners[0].config.ddBaseURL | string | `"https://app.datadoghq.com"` | The base URL of the Datadog |
> | runners[0].config.modes | list | `["workflowAutomation","appBuilder"]` | Modes that the runner can run in |
> | runners[0].config.privateKey | string | `"PRIVATE_KEY_FROM_CONFIG"` | User to specify the runner's privateKey from the enrollment page |
> | runners[0].config.urn | string | `"URN_FROM_CONFIG"` | User to specify the runner's URN from the enrollment page |
> | runners[0].kubernetesPermissions | list | `[{"apiGroups":[""],"resources":["pods"],"verbs":["list","get"]},{"apiGroups":["apps"],"resources":["deployments"],"verbs":["list","get"]}]` | List of Kubernetes permissions that the Datadog Private Action Runner will have |
> | runners[0].replicas | int | `1` | Number of instances of Datadog Private Action Runner |

Reviewer 2: ```suggestion
| Key                                  | Description                                                                  | Type   | Default                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|--------------------------------------|------------------------------------------------------------------------------|--------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `common.image`                       | Current Datadog Private Action Runner image                                  | string | `"us-east4-docker.pkg.dev/datadog-sandbox/apps-on-prem/onprem-runner:v0.0.1-alpha22"`                                                                                                                                                                                                                                                                                                                                                                      |
| `runners[0]`                         | Name of the Datadog Private Action Runner                                    | object | `{"config":{"actionsAllowlist":["com.datadoghq.kubernetes.core.listPod"],"appBuilder":{"port":9016},"ddBaseURL":"https://app.datadoghq.com","modes":["workflowAutomation","appBuilder"],"privateKey":"PRIVATE_KEY_FROM_CONFIG","urn":"URN_FROM_CONFIG"},"kubernetesPermissions":[{"apiGroups":[""],"resources":["pods"],"verbs":["list","get"]},{"apiGroups":["apps"],"resources":["deployments"],"verbs":["list","get"]}],"name":"default","replicas":1}` |
| `runners[0].config`                  | Configuration for the Datadog Private Action Runner                          | object | `{"actionsAllowlist":["com.datadoghq.kubernetes.core.listPod"],"appBuilder":{"port":9016},"ddBaseURL":"https://app.datadoghq.com","modes":["workflowAutomation","appBuilder"],"privateKey":"PRIVATE_KEY_FROM_CONFIG","urn":"URN_FROM_CONFIG"}`                                                                                                                                                                                                             |
| `runners[0].config.actionsAllowlist` | List of actions that the Datadog Private Action Runner is allowed to execute | list   | `["com.datadoghq.kubernetes.core.listPod"]`                                                                                                                                                                                                                                                                                                                                                                                                                |
| `runners[0].config.appBuilder.port`  | Required port for App Builder Mode                                           | int    | `9016`                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `runners[0].config.ddBaseURL`        | The base URL of the Datadog app                                              | string | `"https://app.datadoghq.com"`                                                                                                                                                                                                                                                                                                                                                                                                                              |
| `runners[0].config.modes`            | Modes that the runner can run in                                             | list   | `["workflowAutomation","appBuilder"]`                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `runners[0].config.privateKey`       | The runner's privateKey from the enrollment page                             | string | `"PRIVATE_KEY_FROM_CONFIG"`                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `runners[0].config.urn`              | The runner's URN from the enrollment page                                    | string | `"URN_FROM_CONFIG"`                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| `runners[0].kubernetesPermissions`   | List of Kubernetes permissions that the Datadog Private Action Runner has    | list   | `[{"apiGroups":[""],"resources":["pods"],"verbs":["list","get"]},{"apiGroups":["apps"],"resources":["deployments"],"verbs":["list","get"]}]`                                                                                                                                                                                                                                                                                                               |
| `runners[0].replicas`                | Number of instances of the Datadog Private Action Runner                     | int    | `1`                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
```

Reviewer 2: Also, is that really the correct default value for "Name of the Datadog Private Action Runner"?

Author: the comments came through really messy, I fixed them so now they're much more readable!
